### PR TITLE
Variant refactor

### PR DESCRIPTION
**Pull request by cpacia**

### CODE REVIEW

File: repo/db/inventory.go [hunk 1]
```diff
@@ -42,7 +42,7 @@ func (i *InventoryDB) Get(slug string) (map[string]int, error) {
 	i.lock.RLock()
 	defer i.lock.RUnlock()
 	ret := make(map[string]int)
-	stm := `select * from inventory where slug like "` + slug + `%";`
+	stm := `select * from inventory where slug="` + slug + `";`
```
Comment (tyler-smith) [inline repo/db/inventory.go]: As long as this function is called without user input it should be safe but I think we should always use parameterization for stuff like this so somebody down the line doesn't pass in user input without realizing how dangerous it is.

File: repo/db/inventory.go [hunk 2]
```diff
@@ -70,23 +72,31 @@ func (i *InventoryDB) GetAll() (map[string]int, error) {
...
-	_, err := i.db.Exec("delete from inventory where slug=?", path)
+	_, err := i.db.Exec("delete from inventory where slug=? and variant=?", slug, variant)
 	return err
 }
 
 func (i *InventoryDB) DeleteAll(slug string) error {
 	i.lock.Lock()
 	defer i.lock.Unlock()
-	stm := `delete from inventory where slug like "` + slug + `%";`
+	stm := `delete from inventory where slug="` + slug + `";`
```
Comment (tyler-smith) [inline repo/db/inventory.go]: This should also be parameterized

File: api/jsonapi.go [hunk 3]
```diff
@@ -1016,7 +1028,12 @@ func (i *jsonAPIHandler) POSTInventory(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 	for _, in := range invList {
-		err := i.node.Datastore.Inventory().Put(in.Slug, in.Quantity)
+		formatted, err := json.Marshal(in.Variants)
```
Comment (tyler-smith) [inline api/jsonapi.go]: Why store things as opaque blobs instead of normalized tables? It prevents atomicity, queryability, and data integrity.

**Comment by JustinDrake** [comment]
What design did you end up picking @cpacia? Would it be worth having a spec to limit the reverse-engineering effort for code reviewers? Would you care updating #368?

Briefly looking at [the protobuf definition](https://github.com/OpenBazaar/openbazaar-go/blob/8182617ef68ab549fb89e64ca217239be3814f18/pb/protos/contracts.proto#L65), the `surcharge` feels unnatural:

1. Why use `surchage` instead of absolute prices? `surchage` shapes the thinking to a specific pricing scheme. Surchages can be the pricing metaphor openbazaar-desktop adopts without imposing it to the protocol.
2. Is it a percentage surcharge? Is it a absolute surcharge?
3. How are rounding errors handled? If percentage surcharges are used, rounding issues are a can of worms.

What do @tyler-smith and @jjeffryes think of absolute prices versus surcharges at the protocol level?

**Comment by cpacia** [comment]
> Why use surchage instead of absolute prices? surchage shapes the thinking to a specific pricing scheme. Surchages can be the pricing metaphor openbazaar-desktop adopts without imposing it to the protocol.

It's added (or subtracted) from the base price of the listing. It would be weird if it was the total because then the base price of the item would not be used for anything. 

> Is it a percentage surcharge? Is it a absolute surcharge?
Absolute value

> How are rounding errors handled? If percentage surcharges are used, rounding issues are a can of worms.

Everything is converted into satoshis when calculating the total price. This is the same as it was before, just with a different schema.

**Comment by JustinDrake** [comment]
> It would be weird if it was the total because then the base price of the item would not be used for anything.

The base price can be removed, completely. As you noticed, having `n+1` amounts to describe `n` prices is redundant. Using `n` prices to describes `n` prices seems the most natural.

**Comment by cpacia** [comment]
The issue is we need a single price to display in the UI. It's probably not appropriate to randomly pick the /small/blue/cotton/ variant and use that as the display price.

**Comment by JustinDrake** [comment]
> It's probably not appropriate to randomly pick 

Sure, a random pick is wrong. The default Sku could be implicit by using the first Sku in the `repeated` list. (The list is ordered, right?)

**Pull request merged**